### DISCUSSION TITLE
Compiling a native C module on ESP-WROOM-32

### DISCUSSION THREAD

**Discussion by IcedRooibos** [micropython/micropython#11125] [ESP32]
I'm trying to get started with compiling a native module on ESP32, I am using the ESP-WROOM-32 (DO1T ESP32 DEVKIT V1).

I have my architecture in the Makefile set to `xtensawin` and it compiles fine but when I load it onto the device and try and import it, I get an error `ValueError: incompatible .mpy file`

For now I'm just using the factorial.c code from https://docs.micropython.org/en/kookaberry/develop/natmod.html without any changes.

I've previously built micropython from source using the same setup and tested it and it works fine so I am thinking I have misconfigured something in the Makefile for the factorial module. Any thoughts on where I might have gone wrong? 

In case there is any information I should be including here that I haven't please do let me know, I will add it in.

For the sake of completeness:

Makefile
```
# Location of top-level MicroPython directory
MPY_DIR = /home/IcedRooibos/micropython

# Name of module
MOD = factorial

# Source files (.c or .py)
SRC = factorial.c

# Architecture to build for (x86, x64, armv6m, armv7m, xtensa, xtensawin)
ARCH = xtensawin

# Include to get the rules for compiling and linking the module
include $(MPY_DIR)/py/dynruntime.mk

```

factorial.c
```
// Include the header file to get access to the MicroPython API
#include "py/dynruntime.h"

// Helper function to compute factorial
STATIC mp_int_t factorial_helper(mp_int_t x) {
    if (x == 0) {
        return 1;
    }
    return x * factorial_helper(x - 1);
}

// This is the function which will be called from Python, as factorial(x)
STATIC mp_obj_t factorial(mp_obj_t x_obj) {
    // Extract the integer from the MicroPython input object
    mp_int_t x = mp_obj_get_int(x_obj);
    // Calculate the factorial
    mp_int_t result = factorial_helper(x);
    // Convert the result to a MicroPython integer object and return it
    return mp_obj_new_int(result);
}
// Define a Python reference to the function above
STATIC MP_DEFINE_CONST_FUN_OBJ_1(factorial_obj, factorial);

// This is the entry point and is called when the module is imported
mp_obj_t mpy_init(mp_obj_fun_bc_t *self, size_t n_args, size_t n_kw, mp_obj_t *args) {
    // This must be first, it sets up the globals dict and other things
    MP_DYNRUNTIME_INIT_ENTRY

    // Make the function available in the module's namespace
    mp_store_global(MP_QSTR_factorial, MP_OBJ_FROM_PTR(&factorial_obj));

    // This must be last, it restores the globals dict
    MP_DYNRUNTIME_INIT_EXIT
}

```

**Comment by IcedRooibos** [comment]
I was initially uploading the file in `factorial/factorial.mpy` but instead maybe I should have uploaded the file in `factorial/build/factorial.native.mpy`. I renamed it to `factorial.mpy` before uploading (optional but just makes more sense to do it).

Now I can import factorial but I can't run factorial.factorial

```
>>> import factorial
>>> factorial.factorial(10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'factorial'
```

Seems somewhere the name factorial hasn't registered correctly?

```
>>> dir(factorial)
['__class__', '__name__']
>>> factorial.__class__
<class 'module'>
```

I then also tried building `micropython/examples/natmod/features0` https://github.com/micropython/micropython/tree/master/examples/natmod/features0 without any changes except architecture and similar issue

```
>>> features0.factorial(10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'module' object has no attribute 'factorial'
```

Out of curiosity does anyone know what is the .mpy file in `factorial/factorial.mpy` and how is it different to `factorial.native.mpy`?

Edit: Quite weird, coming back to this now it seems to work. Not sure what to suggest except perhaps a restart after installing the idf and remember to source ./export.sh

**Reply by dpgeorge** [reply]
> Out of curiosity does anyone know what is the .mpy file in `factorial/factorial.mpy` and how is it different to `factorial.native.mpy`?

They are both the same if you don't have any .py files that are merged into the native .mpy file.  If you do have .py files (see the `examples/natmod/features2` example) then the `factorial/factorial.mpy` file is the one that has all code merged together.

You should always use the `factorial/factorial.mpy` file (or whatever it's called for your module).

**Comment by damiencorpataux** [comment]
I had the exact same problem here and found a solution.

**My mistake:** I compiled the *native C module* `factorial.mpy` using micropython 1.20.0 (and esp-idf 5.0.2), but I was running it on an ESP-WROOM-32 with micropython 1.19.x.

**Solution:** I flashed micropython 1.20.0 on the ESP-WROOM-32 and it works.

**Details:**

Using micropython 1.20.0:
```
$ git --git-dir toolchain/micropython/.git log -1
```
```
commit 294baf52b346e400e2255c6c1e82af5b978b18f7 (HEAD -> master, tag: v1.20.0)
Author: Damien George <damien@micropython.org>
Date:   Wed Apr 26 15:42:28 2023 +1000

    all: Bump version to 1.20.0.
```

Using esp-idf 5.0.2:   
```
$ git --git-dir toolchain/esp-idf/.git log -1
```
```
commit 5181de8ac5ec5e18f04f634da8ce173b7ef5ab73 (HEAD, tag: v5.0.2)
Author: Ivan Grokhotkov <ivan@espressif.com>
Date:   Wed May 10 09:59:18 2023 +0200

    versions: Update version to 5.0.2
```

Compiling native C module:
```
$ make -B
```
```
GEN build/factorial.config.h
CC factorial/factorial.c
LINK build/factorial/factorial.o
arch:         EM_XTENSA
text size:    101
rodata size:  8
bss size:     0
GOT entries:  3
GEN factorial.mpy
```

Executing on ESP-WROOM-32:
```
$ mpytool -p /dev/ttyUSB0 put factorial.mpy && mpytool -p /dev/ttyUSB0 repl
```
```
Entering REPL mode, to exit press CTRL + ]

$ mpytool -p /dev/ttyUSB0 put factorial.mpy && mpytool -p /dev/ttyUSB0 repl
Entering REPL mode, to exit press CTRL + ]

>>> import factorial
>>> factorial.factorial(4)
24
>>> 
```

HTH

**Comment by jonnor** [comment]
Mismatched versions of MicroPython for the firmware and what the module is build against by far the most likely cause of this.
